db/range_del_aggregator.cc

```diff
@@ -9,233 +9,146 @@
-      } else {
-        InvalidateTombstoneMapPositions();
-      }
-      first_iter = false;
+  bool ShouldDelete(const ParsedInternalKey& parsed,
+                    RangeDelPositioningMode mode) {
+    if (iter_ == rep_.end() &&
+        (mode == RangeDelPositioningMode::kForwardTraversal ||
+         mode == RangeDelPositioningMode::kBackwardTraversal)) {
+      // invalid (e.g., if AddTombstones() changed the deletions), so need to
+      // reseek
+      mode = RangeDelPositioningMode::kBinarySearch;
     }
-    ParsedInternalKey parsed_key;
-    if (!ParseInternalKey(input->key(), &parsed_key)) {
-      return Status::Corruption("Unable to parse range tombstone InternalKey");
+    switch (mode) {
+      case RangeDelPositioningMode::kFullScan:
+        assert(false);
+      case RangeDelPositioningMode::kForwardTraversal:
+        assert(iter_ != rep_.end());
+        if (iter_ == rep_.begin() &&
+            ucmp_->Compare(parsed.user_key, iter_->first) < 0) {
+          // before start of deletion intervals
+          return false;
+        }
+        while (std::next(iter_) != rep_.end() &&
+               ucmp_->Compare(std::next(iter_)->first, parsed.user_key) <= 0) {
+          ++iter_;
+        }
+        break;
+      case RangeDelPositioningMode::kBackwardTraversal:
+        assert(iter_ != rep_.end());
+        while (iter_ != rep_.begin() &&
+               ucmp_->Compare(parsed.user_key, iter_->first) < 0) {
+          --iter_;
+        }
+        if (iter_ == rep_.begin() &&
+            ucmp_->Compare(parsed.user_key, iter_->first) < 0) {
+          // before start of deletion intervals
+          return false;
+        }
+        break;
+      case RangeDelPositioningMode::kBinarySearch:
+        iter_ = rep_.upper_bound(parsed.user_key);
+        if (iter_ == rep_.begin()) {
+          // before start of deletion intervals
+          return false;
+        }
+        --iter_;
+        break;
     }
-    RangeTombstone tombstone(parsed_key, input->value());
-    AddTombstone(std::move(tombstone));
-    input->Next();
+    assert(iter_ != rep_.end() &&
+           ucmp_->Compare(iter_->first, parsed.user_key) <= 0);
+    assert(std::next(iter_) == rep_.end() ||
+           ucmp_->Compare(parsed.user_key, std::next(iter_)->first) < 0);
+    return parsed.sequence < iter_->second.seq_;
   }
-  if (!first_iter) {
-    rep_->pinned_iters_mgr_.PinIterator(input.release(), false /* arena */);
-  }
-  return Status::OK();
-}
 
-void RangeDelAggregator::InvalidateTombstoneMapPositions() {
-  if (rep_ == nullptr) {
-    return;
-  }
-  for (auto stripe_map_iter = rep_->stripe_map_.begin();
-       stripe_map_iter != rep_->stripe_map_.end(); ++stripe_map_iter) {
-    stripe_map_iter->second.iter = stripe_map_iter->second.raw_map.end();
+  bool IsRangeOverlapped(const Slice&, const Slice&) {
+    // so far only implemented for non-collapsed mode since file ingestion (only
+    // client) doesn't use collapsing
+    assert(false);
```

Comment:
This assert only fires in debug builds. Should this instead be something like `fprintf(stderr, ...); abort()`?

Reply:
Done.